packages/api-contract/src/ContractRegistry.ts

```diff
@@ -0,0 +1,170 @@
+    });
+  }
+
+  public validateMethods ({ contract: { messages } }: t.ContractABIPre): void {
+    messages.forEach((method): void => {
+      const unknownKeys = Object.keys(method).filter((key): boolean => !['args', 'docs', 'mutates', 'name', 'selector', 'return_type'].includes(key));
+
+      assert(unknownKeys.length === 0, `Unknown keys ${unknownKeys.join(', ')} found in ABI args for messages.${method.name}`);
+      const { name, selector, return_type: returnType } = method;
+      assert(isNumber(name) && isString(this.stringAt(name)), `Expected name for messages.${method.name}`);
+      assert(isNumber(selector), `Expected selector for messages.${method.name}`);
+      assert(isNull(returnType) || (isNumber(returnType.ty) && isObject(this.typeDefAt(returnType.ty))), `Expected return_type for messages.${method.name}`);
+
+      this.validateArgs(`messages.${method.name}`, method.args);
+    });
+  }
+
+  public validateAbi (abi: t.ContractABIPre): void {
+    const unknownKeys = Object.keys(abi.contract).filter((key): boolean => !['constructors', 'docs', 'events', 'messages', 'name'].includes(key));
+    assert(unknownKeys.length === 0, `Unknown fields ${unknownKeys.join(', ')} found in ABI`);
+
+    const { contract: { constructors, messages, name } } = abi;
+    assert(constructors && messages && isString(this.stringAt(name)), 'ABI should have constructors, messages & name sections');
+
+    this.validateConstructors(abi);
+    this.validateMethods(abi);
+  }
+
+  public createMethod (name: string, method: Partial<t.ContractABIMethod> & t.ContractABIMethodBase): t.ContractABIFn {
+    const args = method.args.map(({ name, type }): t.ContractABIFnArg => {
+      assert(isObject(type), `Invalid type at index ${type}`);
+      return {
+        name: stringCamelCase(name),
+        type
+      };
+    });
+    const Clazz = createArgClass(args, isUndefined(method.selector) ? {} : { __selector: 'u32' });
+    const baseStruct: { [index: string]: any } = { __selector: method.selector };
+    const encoder = (...params: CodecArg[]): Uint8Array => {
+      assert(params.length === args.length, `Expected ${args.length} arguments to contract ${name}, found ${params.length}`);
+
+      const u8a = new Clazz(
+        args.reduce((mapped, { name }, index): Record<string, CodecArg> => {
+          mapped[name] = params[index];
+
+          return mapped;
+        }, { ...baseStruct })
+      ).toU8a();
+
+      return Compact.addLengthPrefix(u8a);
+    };
+
+    const fn = (encoder as t.ContractABIFn);
+
+    fn.args = args;
+    fn.isConstant = !(method.mutates || false);
+    fn.type = method.returnType || null;
+
+    return fn;
+  }
+
+  public convertAbi ({ contract, storage }: t.ContractABIPre): t.ContractABI {
+    return {
+      contract: this.convertContract(contract),
+      storage: this.convertStorage(storage)
+    };
+  }
+
+  public convertType ({ ty, display_name: displayNameIndices }: t.ContractABITypePre): TypeDef {
```

Comment:
While scanning through quickly, I had a few of those moments where i needed to take a couple of minutes to understand what a function was doing.

Some small comments on functions with not-explicit names would be much appreciated!